text/0000-feature-metadata.md

```diff
@@ -0,0 +1,309 @@
+exists. Semantics will remain unchanged.
+
+This is a required key.
+
+## `doc`
+
+`doc` is the most straightforward: it accepts markdown-flavored text, and should
+be thought of as the equivalent to a `#[doc(...)]` attribute. Like doc comments,
+the first line should be usable as a summary, though most use cases will only
+have one line.
+
+There is nothing in this RFC that cargo `must` do with this action, since it is
+mainly intended for the consumption of `rustdoc` or `docs.rs`. However, it can
+be used for general diagnostic information.
+
+## `deprecated`
+
+`deprecated` should be thought of as the equivalent of the [`deprecated`]
+attribute in Rust source. The value can be a boolean, string, or an object with
+`since` or `note` keys. Schema rules are as follows:
+
+-   If a boolean value, `false` indicates not deprecated and `true` indicates
+    deprecated
+-   If an object, the keys `since` and/or `note` can be specified
+-   An empty object is not allowed to avoid the ambiguous
+    `foo = { deprecated = {} }`
+-   A string `foo = { deprecated = "my msg" }` will be equivalent to if that string
+    was specified in the `note` field:
+    `foo = { deprecated = { note = "my msg" } }`
+-   If not specified, the default is `false`
+
+If a downstream crate attempts to use a feature marked `deprecated`, Cargo
+should produce a warning. There are two exceptions to this:
+
+-   If a `since` is specified that is later than the crate's current version,
+    this warning should not be emitted.
+-   This warning should not be emitted for crates that reexport the feature
+    under a feature also marked deprecated. For example: crate `foo` exports
+    feature `phooey`, and crate `bar` exports feature `barred = ["foo/phooey"]`.
+    If `foo` markes `bar` as deprecated, checking `bar` will emit a warning
+    unless `barred` is also marked `deprecated.
+
+## `public`
+
+`public` is a boolean value that defaults to `true`. It can be thought of as
+`pub` in Rust source files, with the exception of being true by default. If set
+to `false`, Cargo should forbid its use with an error message on any downstream
+crates.
+
+In general, marking a feature `public = false` should make tooling treat the
+feature as non-public API. That includes:
+
+-   The feature should not be accepted by `cargo add --features`
+-   The feature should not be reported from the feature output report of `cargo
+add`
+-   In the future, `rustdoc` should not document these features unless
+    `--document-private-items` is specified
+-   A future tool like `cargo info` shouldn't display information about these
+    features
+
+There needs to be an escape hatch for this for things like benchmarks - RFC TBD
+on how this works.
+
+## `unstable`
+
+`unstable` is a boolean value that defaults to `false`. It should indicate that
+a feature gates something that is not API-stable, or can only be built using
+`nightly` Rust.
```

Comment:
I feel like these are two separate concerns:

- For API-unstable features, there's no need to require a Nightly compiler.
  Instead, it would make more sense to require a `--unstable-features` flag, a `CARGO_UNSTABLE_FEATURES=1` env variable, or require the version to be pinned, so `cargo update` can't break the build.

- For features that only work on Nightly Rust, `nightly = true` would be a more descriptive field. However, a more flexible solution would to be a `rust-version = "<channel>"` field; for example, `rust-version = "nightly"` means that the feature can only be enabled in the Nightly channel. `rust-version = "1.55.0"` means that the feature has a MSRV of 1.55.0, similarly to [`package.rust-version`](https://doc.rust-lang.org/cargo/reference/manifest.html#the-rust-version-field), but scoped to a particular feature.

Reply:
While these are separate use cases, `unstable` can be used to help with the "nightly" feature workflow, even if it isn't the most ideal.

Reply:
As I mentioned [here](https://github.com/rust-lang/rfcs/pull/3416#discussion_r1172915429), I think I'm just going to drop the `unstable` flag from this RFC. The best way forward isn't exactly clear to me, and I think the semantics are a bit more debatable than the rest of the flags.

It's definitely still worth discussing still though, just to have an idea what future plans may look like. I'll write any specific goals up under future possibilities.

Reply:
I did remove `unstable` but think there will be more to discuss here at some point, so I added the `rust-version` suggestion under future possibilities: https://github.com/rust-lang/rfcs/pull/3416/files#diff-b1b8fe9a7278ab777a7fa13af204b0717b1ab818716d76854ed9617569ce08d9R284-R286